core/container.go

```diff
@@ -339,6 +339,54 @@ func (container *Container) WithFS(ctx context.Context, dir *Directory) (*Contai
 	return &Container{ID: id}, nil
 }
 
+func (container *Container) WithDirectory(ctx context.Context, subdir string, src *Directory, filter CopyFilter) (*Container, error) {
+	payload, err := container.ID.decode()
+	if err != nil {
+		return nil, err
+	}
+	fs, err := container.FS(ctx)
```

Comment:
For this and the other new APIs, I don't think this will account for mounts. The `locatePath` func in this file has the logic for finding the right mount

Reply:
You mean writing in the mount if it's a mounted directory?

Reply:
Not sure I see how that would work.  `Directory` and `File` are read operations, here it's a write operation. How can we update the `*Directory` returned from `locatePath`? If a mount is returned what happens?

Reply:
In the current change we get the rootfs, add the directory to it and then return the container with that new rootfs set. To handle mounts, we determine which mount the path points to, add the directory to it, and then return the container with that new mount at its mount point.

From the LLB perspective, both the rootfs and any other normal (i.e. not cache/tmpfs) mounts are more or less the same, they are just states. So updating the rootfs isn't really much different than updating a mount at somewhere besides "/".

Reply:
To be clear, I think `locatePath` might need some minor refactoring to support this, but we need essentially the same logic as is present in there today for mapping a path -> llb state for mount at that path.

Reply:
Just pushed an update, updated only `WithDirectory` to give it a try.

I'm like 70% confident at most that I'm doing the right thing. There's so many paths involved: the workdir, the mount point, the relative path of the directory within the mount point, etc.

And the directory itself has a path.

For instance I tried to use `container.WithMountedDirectory` but that uses `payload.Dir` which is set by `locatePath` to be a relative path within the mount.

Reply:
I 100% agree this is all extremely confusing. For better or worse though, I think it's all just inherent to the problem rather than indicative of us doing something wrong though. IME basically anything involving filepath calculations ends up being 100x more complicated.

For the merge+diff stuff in buildkit (which also fell into complicated filepath logic) I just compensated by having enormous numbers of test cases. I think that's probably the best approach here too (and we have been doing that so far).

If there was a library that could do all of this for us, that'd be great, but I suspect the particulars of our use case are too specific (would be delighted to be wrong though).

Either way, commenting on the code with one *possible* bug I noticed.